map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,172 @@
+
+  if (gpsInfo1.HasSpeed() && gpsInfo2.HasSpeed())
+    extrapolated.m_speed = extrapolator.Extrapolate(gpsInfo1.m_speed, gpsInfo2.m_speed);
+
+  return extrapolated;
+}
+
+// Extrapolator::Routine ---------------------------------------------------------------------------
+Extrapolator::Routine::Routine(ExtrapolatedLocationUpdate const & update)
+  : m_extrapolatedLocationUpdate(update)
+{
+}
+
+void Extrapolator::Routine::Do()
+{
+  while (!IsCancelled())
+  {
+    {
+      GetPlatform().RunTask(Platform::Thread::Gui, [this]() {
+        lock_guard<mutex> guard(m_mutex);
+        uint64_t const extrapolationTimeMs = kExtrapolationPeriodMs * m_extrapolationCounter;
+        if (extrapolationTimeMs >= kMaxExtrapolationTimeMs)
+          return;
+
+        if (DoesExtrapolationWork(extrapolationTimeMs))
+        {
+          location::GpsInfo gpsInfo =
+              LinearExtrapolation(m_beforeLastGpsInfo, m_lastGpsInfo, extrapolationTimeMs);
+          m_extrapolatedLocationUpdate(gpsInfo);
+        }
+        else
+        {
+          if (m_lastGpsInfo.m_source != location::EUndefined)
+          {
+            location::GpsInfo gpsInfo = m_lastGpsInfo;
+            m_extrapolatedLocationUpdate(gpsInfo);
+          }
+        }
+      });
+
+      lock_guard<mutex> guard(m_mutex);
+      if (m_extrapolationCounter != m_extrapolationCounterUndefined)
+        ++m_extrapolationCounter;
+    }
+    // @TODO(bykoinako) Method m_extrapolatedLocationUpdate() is run on gui thread every
+    // |kExtrapolationPeriodMs| milliseconds. But after changing GPS position
+    // (that means after a call of method Routine::SetGpsInfo())
+    // m_extrapolatedLocationUpdate() should be run immediately on gui thread.
+    this_thread::sleep_for(std::chrono::milliseconds(kExtrapolationPeriodMs));
+  }
+}
+
+void Extrapolator::Routine::SetGpsInfo(location::GpsInfo const & gpsInfo)
+{
+  lock_guard<mutex> guard(m_mutex);
+  m_beforeLastGpsInfo = m_lastGpsInfo;
+  m_lastGpsInfo = gpsInfo;
+  m_extrapolationCounter = 0;
+}
+
+bool Extrapolator::Routine::DoesExtrapolationWork(uint64_t extrapolationTimeMs) const
+{
+  // Note. It's possible that m_beforeLastGpsInfo.m_timestamp >= m_lastGpsInfo.m_timestamp.
+  // It may happen in rare cases because GpsInfo::m_timestamp is not monotonic generally.
+  // Please see comment in declaration of class GpsInfo for details.
+
+  if (m_extrapolationCounter == m_extrapolationCounterUndefined ||
+      m_lastGpsInfo.m_source == location::EUndefined ||
+      m_beforeLastGpsInfo.m_source == location::EUndefined ||
+      m_beforeLastGpsInfo.m_timestamp >= m_lastGpsInfo.m_timestamp)
+  {
+    return false;
+  }
+
+  double const distM =
+      ms::DistanceOnEarth(m_beforeLastGpsInfo.m_latitude, m_beforeLastGpsInfo.m_longitude,
+                          m_lastGpsInfo.m_latitude, m_lastGpsInfo.m_longitude);
+  double const timeS = m_lastGpsInfo.m_timestamp - m_beforeLastGpsInfo.m_timestamp;
```

Comment:
In what units timestamp is measured?

Reply:
In seconds:
```
  class GpsInfo
  {
  public:
    TLocationSource m_source = EUndefined;
    /// \note |m_timestamp| is calculated based on platform methods which don't
    /// guarantee that |m_timestamp| is monotonic. |m_monotonicTimeMs| should be added to
    /// class |GpsInfo|. This time should be calculated based on Location::getElapsedRealtimeNanos()
    /// method in case of Android. How to calculate such time in case of iOS should be
    /// investigated.
    /// \note For most cases |m_timestamp| is monotonic.
    double m_timestamp = 0.0;             //!< seconds from 1st Jan 1970
...
```